src/object/srv_obj.c

```diff
@@ -265,6 +281,151 @@ obj_bulk_bypass(d_sg_list_t *sgl, crt_bulk_op_t bulk_op)
 	}
 }
 
+static int
+bulk_transfer_sgl(daos_handle_t ioh, crt_rpc_t *rpc, crt_bulk_t remote_bulk,
+		  off_t remote_off, crt_bulk_op_t bulk_op, bool bulk_bind,
+		  d_sg_list_t *sgl, int sgl_idx, struct obj_bulk_args *p_arg)
+{
+	struct bio_desc		*biod;
+	struct crt_bulk_desc	bulk_desc;
+	crt_bulk_perm_t		bulk_perm;
+	crt_bulk_opid_t		bulk_opid;
+	crt_bulk_t		local_bulk;
+	unsigned int		local_off;
+	unsigned int		iov_idx = 0;
+	size_t			remote_size;
+	int			rc;
+
+	if (remote_bulk == NULL) {
+		D_ERROR("Remote bulk is NULL\n");
+		return -DER_INVAL;
+	}
+
+	rc = crt_bulk_get_len(remote_bulk, &remote_size);
+	if (rc) {
+		D_ERROR("Failed to get remote bulk size "DF_RC"\n", DP_RC(rc));
+		return rc;
+	}
+
+	if (remote_off >= remote_size) {
+		D_ERROR("remote_bulk_off %zu >= remote_bulk_size %zu\n",
+			remote_off, remote_size);
+		return -DER_INVAL;
+	}
+
+	if (daos_io_bypass & IOBP_SRV_BULK) {
+		obj_bulk_bypass(sgl, bulk_op);
+		return 0;
+	}
+
+	biod = daos_handle_is_valid(ioh) ? vos_ioh2desc(ioh) : NULL;
+	bulk_perm = bulk_op == CRT_BULK_PUT ? CRT_BULK_RO : CRT_BULK_RW;
+
+	while (iov_idx < sgl->sg_nr_out) {
+		d_sg_list_t	sgl_sent;
+		size_t		length = 0;
+		unsigned int	start;
+		bool		cached_bulk = false;
+
+		/*
+		 * Skip bulk transfer over IOVs with NULL buffer address,
+		 * these NULL IOVs are 'holes' or deduped records.
+		 */
+		while (iov_idx < sgl->sg_nr_out &&
+		       sgl->sg_iovs[iov_idx].iov_buf == NULL) {
+			remote_off += sgl->sg_iovs[iov_idx].iov_len;
+			iov_idx++;
+		}
+
```

Comment:
i am quite confused about this logic.
Say we have a local sgl that consists of buffers [ iov0, len = 10 ] [ iov1 =NULL, len=5 ] [ iov2, len=20 ]. Assume initial remote_off = 0;
After this loop remote_off will be 5, meaning iov0 in the destination will start at the offset 5 (and might overflow iov space for it); is this really what is intended?

Reply:
In your example, the remote_off is still 0 after the loop since iov0 isn't NULL.

Reply:
why is it still 0? the loop goes through all iov_idx from 0 until sgl->sg_nr_our, so remote_off would get modified as iov[iov_idx=1]  == NULL, no?

Reply:
nvm, brain freeze on my part